whois-nrtm4/src/main/java/net/ripe/db/nrtm4/NrtmFileService.java

```diff
@@ -0,0 +1,115 @@
+import net.ripe.db.nrtm4.persist.NrtmVersionInfo;
+import net.ripe.db.nrtm4.persist.NrtmVersionInfoRepository;
+import net.ripe.db.nrtm4.persist.SnapshotFile;
+import net.ripe.db.nrtm4.persist.SnapshotFileRepository;
+import net.ripe.db.nrtm4.publish.PublishableSnapshotFile;
+import net.ripe.db.nrtm4.publish.SnapshotFileStreamer;
+import org.springframework.stereotype.Service;
+
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+
+import static net.ripe.db.nrtm4.NrtmConstants.DELTA_PREFIX;
+import static net.ripe.db.nrtm4.NrtmConstants.NOTIFICATION_PREFIX;
+import static net.ripe.db.nrtm4.NrtmConstants.SNAPSHOT_PREFIX;
+
+
+@Service
+public class NrtmFileService {
+
+    private final NotificationFileRepository notificationFileRepository;
+    private final NrtmFileRepo nrtmFileRepo;
+    private final NrtmVersionInfoRepository nrtmVersionInfoRepository;
+    private final DeltaFileRepository deltaFileRepository;
+    private final SnapshotFileRepository snapshotFileRepository;
+    private final SnapshotFileStreamer snapshotFileStreamer;
+    private final Monitor deltaMutex = new Monitor();
+    private final Monitor snapshotMutex = new Monitor();
+
+    NrtmFileService(
+        final NotificationFileRepository notificationFileRepository,
+        final NrtmFileRepo nrtmFileRepo,
+        final NrtmVersionInfoRepository nrtmVersionInfoRepository,
+        final DeltaFileRepository deltaFileRepository,
+        final SnapshotFileRepository snapshotFileRepository,
+        final SnapshotFileStreamer snapshotFileStreamer
+    ) {
+        this.notificationFileRepository = notificationFileRepository;
+        this.nrtmFileRepo = nrtmFileRepo;
+        this.nrtmVersionInfoRepository = nrtmVersionInfoRepository;
+        this.deltaFileRepository = deltaFileRepository;
+        this.snapshotFileRepository = snapshotFileRepository;
+        this.snapshotFileStreamer = snapshotFileStreamer;
+    }
+
+    void writeFileToStream(final String name, final OutputStream out) throws IOException {
+        if (name == null || name.length() > 256) {
+            throw new IllegalArgumentException("Invalid NRTM file name");
+        }
+        if (name.startsWith(NOTIFICATION_PREFIX)) {
+            // TODO: how do we know which source to use when serving a notification request?
+            final NotificationFile notificationFile = notificationFileRepository.getNotificationFile(1);
+            out.write(notificationFile.getPayload().getBytes(StandardCharsets.UTF_8));
+            return;
+        }
+        syncNrtmFileToFileSystem(name);
+        nrtmFileRepo.streamFromFile(name, out);
+    }
+
+    void syncNrtmFileToFileSystem(final String name) throws IOException {
+        if (name.startsWith(DELTA_PREFIX)) {
+            if (!nrtmFileRepo.checkIfFileExists(name)) {
```

Comment:
Can we simplify this logic? may be create two separate methods? There are too many if else conditions

Reply:
split into two services and uses the 'return early' principle to avoid nesting